src/host/enclave.h

```diff
@@ -46,49 +78,86 @@ namespace host
   class Enclave
   {
   private:
-    bool is_virtual_enclave;
-
-    oe_enclave_t* e;
+#ifdef CCHOST_SUPPORTS_SGX
+    oe_enclave_t* sgx_handle = nullptr;
+#endif
+#ifdef CCHOST_SUPPORTS_VIRTUAL
+    void* virtual_handle = nullptr;
+#endif
 
   public:
     /**
      * Create an uninitialized enclave hosting the given library.
      *
      * @param path Path to signed enclave library file
-     * @param flags Flags passed to oe_create_enclave. eg OE_ENCLAVE_FLAG_DEBUG,
-     * OE_ENCLAVE_FLAG_SIMULATE. Alternatively, ENCLAVE_FLAG_VIRTUAL will not
-     * use OE at all, instead loading a shared library directly
+     * @param type Type of enclave to load, influencing what flags should be
+     * passed to OE, or whether to dlload a virtual enclave
      */
-    Enclave(const std::string& path, uint32_t flags) :
-      is_virtual_enclave(false),
-      e(nullptr)
+    Enclave(const std::string& path, EnclaveType type)
```

Comment:
So this function does now have more awkward macro-enabled blocks, but the result is a consistent, user-optimised error reporting flow:

- Does the file exist? If not, throw an error
- Is this a type of file that this build of `cchost` can handle? If not, throw an error
- Does this file have the expected suffix? If not, throw an error
- Otherwise try to load it, and report if _that_ fails

Previously steps 2 and 3 were swapped, so you could (if you really tried) get a "have you tried *.enclave.so.signed?" message, spend a while building/finding that image, and then be told by exactly the same `cchost` "actually, I don't want any files like that".